### PR TITLE
Implement EmailSenderTemplate class as new email template sender

### PR DESCRIPTION
**Pull request by RZR-UA**

### CODE REVIEW

**Comment by RZR-UA** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2380

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java [hunk 1]
```diff
@@ -1,285 +1,95 @@
...
-            Response response = wrapper.send(mail);
-            LOGGER.info("Response code: " + response.getStatusCode());
-            return true;
-        } catch (Exception e) {
-        	LOGGER.warn(e.getMessage(), e);
-        	return false;
-		}
+    public boolean sendPromocodesEmails(String mailTo, String templateId, String promocodes) throws FileNotFoundException, UnsupportedEncodingException {
+	    boolean ok = new EmailSenderTemplate()
+			    .load(templateId) // should be "promocode/ios" or "promocode/anroid"
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:62]: To https://osmand.net/admin/info#giveaway need add tips where to create new template 
( https://github.com/osmandapp/web-server-config/tree/main/templates/email )

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java [hunk 2]
```diff
@@ -1,285 +1,95 @@
...
+			    .to(Arrays.asList(mailTo.split(",")))
+			    .set("PROMOCODE", promocodes)
+			    .send()
+			    .isSuccess();
+	    LOGGER.info("sendPromocodesEmails to: " + mailTo + " (" + ok + ")");
+		return ok;
     }
 	
 	public void sendOsmRecipientsDeleteEmail(String email) {
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:71]: In AdminController.java we have API
@PostMapping(path = {"/delete-email"})

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java [hunk 3]
```diff
@@ -1,285 +1,103 @@
...
-		mail.from = from;
-		try {
-			Response response = wrapper.send(mail);
-			LOGGER.info("Response code: " + response.getStatusCode());
-		} catch (Exception e) {
-			LOGGER.warn(e.getMessage(), e);
-		}
+    public void sendOsmAndCloudPromoEmail(String email, String promo) throws FileNotFoundException, UnsupportedEncodingException {
+		boolean ok = new EmailSenderTemplate()
+				.load("cloud/promo")
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:29]: No need lang here?

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/UserdataService.java [hunk 4]
```diff
@@ -319,8 +319,10 @@ public ResponseEntity<String> webUserActivate(String email, String token, String
         }
         return registerNewDevice(email, token, TOKEN_DEVICE_WEB, encoder.encode(password));
     }
-    
-	public ResponseEntity<String> webUserRegister(@RequestParam(name = "email", required = true) String email) throws IOException {
+
+	public ResponseEntity<String> webUserRegister(@RequestParam(name = "email", required = true) String email,
+	                                              @RequestParam(name = "lang", required = false) String lang)
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/UserdataService.java:324]: Request for Android here https://github.com/osmandapp/OsmAnd/pull/19058/files

File: java-tools/OsmAndServerUtilities/build.gradle [hunk 5]
```diff
@@ -9,6 +9,7 @@ dependencies {
 	implementation 'com.google.code.findbugs:jsr305:3.0.2'
 	implementation 'net.sf.kxml:kxml2:2.1.8'
 	implementation 'org.apache.commons:commons-compress:1.17'
+  implementation 'org.apache.commons:commons-email:1.5'
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServerUtilities/build.gradle:12]: Formatting

File: java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderMain.java [hunk 6]
```diff
@@ -82,12 +71,22 @@ public static void main(String[] args) throws SQLException {
             	}
             } else if (arg.equals("--update_block_list")) {
                 updateBlockList = true;
+			} else if (arg.equals("--update_postfix_bounced")) {
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderMain.java:74]: Formatting

File: java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderMain.java [hunk 7]
```diff
@@ -366,68 +428,100 @@ private static Connection getConnection() {
...
+//    	}
+//    	String mailHsh = mailTo;
+//    	if(mailTo.length() > 5) {
+//    		mailHsh = "....." + mailTo.substring(5);
+//    	}
+//        LOGGER.info("Sending mail to: " + mailHsh);
+//        String userHash;
+//        try {
+//            userHash = URLEncoder.encode(Base64.getEncoder().encodeToString(mailTo.getBytes()), "UTF-8");
+//        } catch (UnsupportedEncodingException e) {
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderMain.java:443]: Just use /**/ for avoid transfer change code ownership

Comment (RZR-UA) [reply java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderMain.java:443]: old method removed

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/UserdataService.java (hunk 4 above)
Comment (RZR-UA) [reply java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/UserdataService.java:324]: android/ios commits are merged

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java (hunk 3 above)
Comment (RZR-UA) [reply java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:29]: lang is unknown for cloud/promo templates, so not used

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java (hunk 1 above)
Comment (RZR-UA) [reply java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:70]: Added tips and link to edit templates (github repo)

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java (hunk 2 above)
Comment (RZR-UA) [reply java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:79]: @PostMapping(path = {"/delete-email"})

This API was used to remove osrm_receipients and is not used anymore.

Talked with EZ and decided to keep old email in the comments.

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java [hunk 8]
```diff
@@ -1,285 +1,103 @@
...
-		Mail mail = new Mail(from, topic, to, content);
-		mail.from = from;
-		try {
-			Response response = wrapper.send(mail);
-			LOGGER.info("Response code: " + response.getStatusCode());
-		} catch (Exception e) {
-			LOGGER.warn(e.getMessage(), e);
-		}
+		throw new IllegalStateException("sendOsmRecipientsDeleteEmail() is obsolete");
+//		LOGGER.info("Sending mail to: " + email);
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:81]: No code in comment!

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java [hunk 9]
```diff
@@ -1,285 +1,103 @@
...
-				LOGGER.warn("SMTP error: " + e.getMessage() + " (" + e.getCause().getMessage() + ")");
-				return fakeSuccess();
-			}
-
-			return new Response(SENDGRID_SUCCESS_STATUS_CODE, "", null);
-		}
+	public static void test(String[] args) throws FileNotFoundException, UnsupportedEncodingException {
+		EmailSenderService sender = new EmailSenderService();
+		String email = "dmarc-reports@osmand.net";
+//		sender.sendOsmRecipientsDeleteEmail(email); // obsolete
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/EmailSenderService.java:17]: No code in comments!

File: java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderTemplate.java [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+
+	new EmailSenderTemplate().template("Subject: test\nHi!").to("test@email").to("other@email").send().isSuccess();
+	new EmailSenderTemplate().load("cloud/web", "en").to("osmand@t-online.de").set("TOKEN", "777").send().isSuccess();
+
+*/
+
+public class EmailSenderTemplate {
+	private static final Log LOG = LogFactory.getLog(EmailSenderTemplate.class);
+
+	public String defaultTemplatesDirectory = "/var/www-download/website/templates/email";
```
Comment (ivanPyrohivskyi) [inline java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderTemplate.java:43]: Must be in application.yml

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java [hunk 11]
```diff
@@ -40,7 +42,7 @@ public void register(PromoCampaignRepository.Promo promo) {
 	}
 
 	@Transactional
-	public ResponseEntity<String> addUser(String name, String email) {
+	public ResponseEntity<String> addUser(String name, String email) throws FileNotFoundException, UnsupportedEncodingException {
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java:45]: Exceptions shuold be RuntimeException by guidelines as we don't need here checked exceptions

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java [hunk 12]
```diff
@@ -99,7 +101,7 @@ private String getLastUsers(PromoCampaignRepository.Promo promoCampaign) {
 		return emails.substring(0, emails.length() - 3);
 	}
 
-	public PromoResponse createPromoSubscription(String email, String key, Date expireTime) {
+	public PromoResponse createPromoSubscription(String email, String key, Date expireTime) throws FileNotFoundException, UnsupportedEncodingException {
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java:104]: Exceptions shuold be RuntimeException by guidelines as we don't need here checked exceptions

File: java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderTemplate.java [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+		return this;
+	}
+
+	// optional set fromName, use before load()
+	public EmailSenderTemplate name(String name) {
+		fromName = name;
+		return this;
+	}
+
+	// optional set subject, use before load()
```
Comment (vshcherb) [inline java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderTemplate.java:167]: Comments to explain code bad practice, it's better to use design patterns / approaches and self explaining code. The best API that you can't use incorrectly

Comment (RZR-UA) [reply java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderTemplate.java]: comments removed, documentation clarified

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java (hunk 11 above)
Comment (RZR-UA) [reply java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java]: now caught locally

**Pull request merged**